src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,147 @@
+            size++;
+            growIfArrayFull();
+            System.arraycopy(elementStore, index, elementStore,
+                    nextIndex(index), calculateElementsToCopy(index));
+            set(value, index);
+        } else {
+            throw createErrorWithIndex(index);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (indexIsLegal(index)) {
+            return elementStore[index];
+        } else {
+            throw createErrorWithIndex(index);
+        }
     }
 
     @Override
-    public void set(T value, int index) {
-
+    public T remove(int index) {
+        T oldValue;
+        if (indexIsLegal(index)) {
+            oldValue = elementStore[index];
+            if (isFull()) {
+                elementStore[index] = null;
+                size--;
+                return oldValue;
+            }
+            System.arraycopy(elementStore, nextIndex(index), elementStore,
+                    index, calculateElementsToCopy(index));
+            elementStore[size] = null;
+            size--;
+            return oldValue;
+        } else {
+            throw createErrorWithIndex(index);
+        }
     }
 
     @Override
-    public T remove(int index) {
-        return null;
+    public T remove(T element) {
+        for (int i = 0; i < size; i++) {
+            if ((elementStore[i] == element)
+                    || (elementStore[i] != null && elementStore[i].equals(element))) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException();
     }
 
     @Override
-    public T remove(T element) {
-        return null;
+    public void set(T value, int index) {
+        if (indexIsLegal(index)) {
+            elementStore[index] = value;
+        } else {
+            throw createErrorWithIndex(index);
+        }
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder("[");
+        sb.append(elementStore[0]);
+        for (int i = 1; i < elementStore.length; i++) {
+            if (i < size) {
+                sb.append(", ").append(elementStore[i]);
+            }
+        }
+        sb.append("]");
+        return sb.toString();
+    }
+
+    private int calculateElementsToCopy(int index) {
+        return size - index;
+    }
+
+    private ArrayListIndexOutOfBoundsException createErrorWithIndex(int index) {
+        return new ArrayListIndexOutOfBoundsException("Index: [" + index
+                + "] is out of bounds for list size [" + size + "]!");
+    }
+
+    private void growIfArrayFull() {
+        if (isFull()) {
```

Comment:
```suggestion
    private void growIfArrayFull() {
        if (size == elementStore.length) {
```
method with one line. Not necessary.

Reply:
This expression is reused several times in the code, so I think it's good not to copy-paste it.